### ISSUE TITLE
User Feedback

### ISSUE THREAD

**Issue by jazz-it** [puxplaying/maus#2]
Hi! I use your script, find it very useful, therefore just wanted to drop a note... btw, I've just mentioned it here:
https://forum.manjaro.org/t/what-are-your-experiences-related-to-stability-of-manjaro-ecosystem/154301/32

Would you be interested to improve your script even further by utilizing the power of `aurphan` tool (since it may help maintaining Manjaro quite a bit)?

EDIT: I'd like to take that back. Just found this project that is probably more appropriate for that feature:
https://github.com/excalibur1234/pacui

**Comment by puxplaying** [comment]
Hi, thanks for the feedback! I do also think that this feature would better fit in PacUI. If you look for more CLI tools, you could take a look at [ToolBox](https://github.com/puxplaying/toolbox) (former [bmenu](https://github.com/Chrysostomus/bmenu)), which also makes use of PacUI. It is available via the `bmenu` package on Manjaro.

**Comment by jazz-it** [comment]
FYI: https://forum.manjaro.org/t/does-it-seem-safe-to-remove-all-of-the-following-orphans/12664/9

**Comment by jazz-it** [comment]
Hi again! :) Is there any reason to downgrade pamac?

```
warning: pamac-cli: downgrading from version 10.1.2-0.1 to version 10.0.6-2
warning: pamac-common: downgrading from version 10.1.2-0.1 to version 10.0.6-2
warning: pamac-gtk: downgrading from version 10.1.2-0.1 to version 10.0.6-2
resolving dependencies...
looking for conflicting packages...

Packages (3) pamac-cli-10.0.6-2  pamac-common-10.0.6-2  pamac-gtk-10.0.6-2

Total Installed Size:  13.34 MiB
Net Upgrade Size:      -0.15 MiB

:: Proceed with installation? [Y/n]
```

**Comment by puxplaying** [comment]
Hi, this was a decision of Manjaro, they have downgraded Pamac on the stable branch. You can always check with [Branch Compare](https://manjaro.org/branch-compare/).
![9](https://user-images.githubusercontent.com/28549766/118992572-0e802480-b985-11eb-870e-731e3286bf9a.png)


My script just checks for Downgrades, hence the prompt you see.
See also this post: https://forum.manjaro.org/t/stable-update-2021-05-19-kernels-nvidia-kde-frameworks-plasma-systemd-libreoffice-kde-gear-mobile-ff-tb/67002/46

**Comment by jazz-it** [comment]
Hi, the recent system upgrade had some serious impact for some of the users who updated their system with `pacman` instead of `pamac`. I understood `sudo pamac upgrade` is the recommended approach, therefore I currently use the following as my upgrade routine:

```
sudo pacman-mirrors -f
sudo journalctl --vacuum-size=50M
sudo paccache -ruk0
sudo pamac upgrade
pip install --upgrade pip
python --version
echo "PythonPATH: $PYTHONPATH"
ls -d /usr/lib/python*
```

Maybe you could also consider switching to using `pamac` as the default within your script, according to the comments here:

https://forum.manjaro.org/t/stable-update-2021-06-14-kernels-browsers-mesa-deepin-systemd-gnome-apps-40-2-pipewire-haskell/70192

**Comment by puxplaying** [comment]
Heya, that is exactly what my script does if Pamac is the only AUR helper that is installed.
In case the update fails via Pacman, Pamac will be used and if that fails too, there is a third routine to get it done ;)

**Comment by jazz-it** [comment]
My point was maybe to go with `pamac` first (to avoid possible similar glitches as mentioned above), that's all I suggested. :)

**Comment by puxplaying** [comment]
I get your point, but let me explain ;)

In the current example `pacman` would fail to apply the update anyway and the user would be prompted by `pamac` to apply the update, so that there wouldn't be any glitches. Pacman reverts everything in case it is unable to apply the update.

You can see it as a double security, if `pacman` fails, nothing will happen and `pamac` kicks in - and if that fails there is a manual third routine.

**Comment by jazz-it** [comment]
I downloaded 1.1GB with the latest update. Does it mean the files would be downloaded twice as well, in case `pacman` fails and `pamac` takes over?

**Comment by puxplaying** [comment]
No, because the downloaded packages are cached in `/var/cache/pacman/pkg`

**Comment by jazz-it** [comment]
Ok, then it will work. :) Sorry for flooding.

**Comment by puxplaying** [comment]
Don't worry, I appreciate the feedback, thanks!

**Comment by jazz-it** [comment]
Hi, quite recently I'm struggling to update the following package with your script: `python2-lxml`

This is the output: https://hastebin.com/rinuwiyere.sql

When I try to update it with `sudo pamac upgrade` it works like a charm.

I guess it's a bug, then?

**Comment by puxplaying** [comment]
Hi, according to the output this is not related to **pamac**. As I mentioned [above](https://github.com/puxplaying/manjaro-update/issues/2#issuecomment-861015157), there is a simple logic in place when different AUR helpers are installed. If the user decides to install a supported AUR helper like in your case **yay**, than this helper will be used to update the AUR and not **pamac**. According to the output yay was not able to build the package because of a _syntax error_, so this should be reported upstream if the error with yay is reproducible.

**Comment by jazz-it** [comment]
Yep, I understood correctly the error has nothing to do with `pamac`, but I understood that if `yay` fails, then `pamac` may take over which didn't happen. That's why I thought your script may correct the final outcome of the updating procedure, but I agree with you: `yay` team should be aware of this issue. Ok, I will post the issue upstream. Thanks!

**Comment by puxplaying** [comment]
> but I understood that if `yay` fails, then `pamac` may take over which didn't happen. That's why I thought your script may correct the final outcome of the updating procedure

Sorry if my explanation was confusing. Pamac will only be used if there is no other AUR helper installed, since this is user preference which will be preferred over default behaviour.

**Comment by jazz-it** [comment]
That was useful. Thanks for clarifying!

EDIT: I use Manjaro KDE which comes with `octopi` that requires `yay`. Maybe it would be better to configure the list of wanted AUR helpers within the script somehow, since for me - `pamac` seems to be the only reliable helper.

**Comment by jazz-it** [comment]
For further reference, I reported the issue with `yay` upstream:

https://github.com/Jguer/yay/issues/1554

**Comment by jazz-it** [comment]
> Looks like the package is trying to build against python3 for some reason. Don't see how `yay` could cause that.

Any ideas?

**Comment by puxplaying** [comment]
> Any ideas?

Well it could also be a packaging issue or a issue in the `setup.py` script of `python2-lxml`.
You could try to build (not update) the package with `yay`, `makepkg` and `pamac` to see if the results are different.



> Maybe it would be better to configure the list of wanted AUR helpers within the script somehow, since for me - `pamac` seems to be the only reliable helper.

When multiple helpers are installed, you can set the wanted AUR helper by adding the variable  `AUR_Helper="pamac"`  on top of the script, right after the Shebang, for Pamac in this case.

**Comment by puxplaying** [comment]
I have updated the script according to your feedback, so that now `pamac` will be used as the first helper even if multiple helpers are installed and also added a `AUR_helper` variable to manually set the preferred helper.

**Comment by jazz-it** [comment]
I usually used the following alias to update my packages:

```
alias up='sudo pacman-mirrors -f && sudo journalctl --vacuum-size=50M && sudo paccache -ruk0 && pamac upgrade && gitup && pip install --upgrade pip && tail -n20 /var/log/pacman.log && echo "Last 10 warnings:" && grep warning /var/log/pacman.log | tail -n10  && echo "Last 10 errors:" && grep Error /var/log/pacman.log | tail -n10 && python --version && echo "PythonPATH: $PYTHONPATH" && ls -d /usr/lib/python*'
```

The more I use it, the more frequently I fail to update some AUR packages, but each time I ran your script, it was always successful. Not quite sure where's the catch, but so far I just wanted to send you this positive feedback - your script should be the default for most users. Thank you!

**Comment by jazz-it** [comment]
Now that Arch blocked `pamac` from `AUR`, maybe it would be good idea to separate a package manager tool for `AUR`, perhaps? Or just switching to `paru` as the default altogether.... meh...

**Comment by puxplaying** [comment]
Well Pamac is Manjaro's baby and also has more advantages than just the use of the AUR, like being able to solve file conflicts etc. I'm pretty sure this is just a temporary situation and the Arch and Manjaro devs are working on a solution.

If you need a different AUR helper in the mean time, feel free to set one via the `AUR_Helper=` variable inside the script. Paru is currently not implemented as it needs more testing on my side, but I’m working on it.

**Comment by jazz-it** [comment]
I've noticed `sudo pacman -Sy meld` in line 366, which could be a typo (since `-Sy` is discouraged and treated as a bad practice). Could you replace it with `sudo pacman -Syu meld`, please?

Also, probably removing `--noconfirm` at lines 196 and 223 would be more safer to use?

**Comment by puxplaying** [comment]
> Could you replace it with `sudo pacman -Syu meld`, please?

Thanks, this is fixed now, not sure how it got there anyway.



> Also, probably removing `--noconfirm` at lines 196 and 223 would be more safer to use?

This would defeat the whole purpose of the script, which is made with minimal user intervention in mind.
Maybe you want to explain what you mean with "safer"? Why would a prompt which you have to answer make anything safer, when the result is that you want to update your system?

**Comment by jazz-it** [comment]
There are two different scenarios I could think of:

1) I could accidentally run `manjaro-update` by pressing "up" > "up" > "up" > "enter", instead ie. "up" > "up" > "enter" in my terminal window. Say I didn't want to run that command at the first place, at least not without making a backup first.

2) In an even rarer situation, I might want to revise the list of packages pending for update, before executing the update itself.

**Comment by puxplaying** [comment]
> 1. I could accidentally run `manjaro-update` by pressing "up" > "up" > "up" > "enter", instead ie. "up" > "up" > "enter" in my terminal window. Say I didn't want to run that command at the first place, at least not without making a backup first.

This script will not run without a sudo password. Press ctrl+c if you accidentally run it to exit. Or use the provided launcher and delete it from your shells history.

> 2\. In an even rarer situation, I might want to revise the list of packages pending for update, before executing the update itself.

Pamac is part of all official Manjaro editions and comes with an update indicator which shows the pending updates, so I don't see a need to review the package list in a script like this. 

You may want to take a look at PacUI's system maintain function for a more guided experience.

**Comment by jazz-it** [comment]
Suddenly I have issues with the script when updating the AUR packages, e.g.:
`cp: cannot create regular file '/var/tmp/pamac/dbs/sync/pamac_aur.db': Permission denied`

However, updating with `pamac upgrade` runs normally.

**Comment by puxplaying** [comment]
Seems like a new **pamac** issue then which I'm unable to reproduce, what happens if you run `pamac update -a` in the terminal, since this is exactly what my script does?

**Comment by jazz-it** [comment]
I will try to run it the next time I run into this and report back here.

**Comment by jazz-it** [comment]
I just encountered the same issue and this time I ran `pamac update -a` right after:

```
$ pamac update -a
Preparing...
Synchronizing package databases...
Checking android-platform dependencies...
cp: cannot create regular file '/var/tmp/pamac/dbs/sync/pamac_aur.db': Permission denied
Error: Failed to prepare transaction: target not found: android-platform

Edit build files : [e] 
Apply transaction ? [e/y/N] e
```

The result is exactly the same. The only difference is that this time I pressed `e` + `Ctrl+X` instead of pressing `y` and this time the package updated successfully. So, must be an issue with `pamac`. I reported it here a month ago, but nobody replied yet:

https://forum.manjaro.org/t/stable-update-2022-06-24-kernels-kde-frameworks-cinnamon-5-4-qt5-gstreamer-haskell-python/114857/28?u=jazz

**Comment by jazz-it** [comment]
```
$ ls -al /var/tmp/pamac/dbs/sync/pamac_aur.db
-rw-rw-rw- 1 root root 2560 Jul 24 23:41 /var/tmp/pamac/dbs/sync/pamac_aur.db

$ sudo chown -R $(whoami):$(whoami) /var/tmp/pamac/
```

That should fix it.

**Comment by FaFre** [comment]
I encounter the same issue and tried your fix with chown @jazz-it 
The result is, my pamac_aur.db got deleted after running `pamac update` with fixed privileges. Pamac doesn't want to create a new database now.

```bash
    ~  ls -al /var/tmp/pamac/dbs/sync/pamac_aur.db                                                                                                                                                1 ✘  46s  
ls: cannot access '/var/tmp/pamac/dbs/sync/pamac_aur.db': No such file or directory
    ~  ls -al /var/tmp/pamac/dbs/sync/                                                                                                                                                                    2 ✘ 
total 73588
drwxrwxrwx 2 fafre fafre     4096  9. Aug 11:31 .
drwxrwxrwx 3 fafre fafre     4096  9. Aug 11:31 ..
-rw-r--r-- 1 fafre fafre  7418408  7. Aug 13:49 community.db
-rw-r--r-- 1 fafre fafre 33921062  7. Aug 13:49 community.files
-rw-r--r-- 1 fafre fafre 20066304 16. Jun 21:40 community.files.part
-rw-r--r-- 1 fafre fafre   169156  7. Aug 13:49 core.db
-rw-r--r-- 1 fafre fafre  1459972  7. Aug 13:49 core.files
-rw-r--r-- 1 fafre fafre  1896112  7. Aug 13:49 extra.db
-rw-r--r-- 1 fafre fafre  9918633  7. Aug 13:49 extra.files
-rw-r--r-- 1 fafre fafre    18496 17. Okt 2021  mhwd.db
-rw-r--r-- 1 fafre fafre   183111  7. Aug 13:49 multilib.db
-rw-r--r-- 1 fafre fafre   267547  7. Aug 13:49 multilib.files
-rw-rw-rw- 1 fafre fafre        0  9. Aug 11:31 refresh_timestamp
    ~  pamac update -a                                                                                                                                                                                      ✔ 
Preparing...
Synchronizing package databases...
Cloning postman-bin build files...
Generating postman-bin information...
Checking postman-bin dependencies...
cp: cannot create regular file '/var/tmp/pamac/dbs/sync/pamac_aur.db': Permission denied
```

Do you have any ideas how to fix this?

**Comment by jazz-it** [comment]
> Do you have any ideas how to fix this?

Did you try: `touch /var/tmp/pamac/dbs/sync/pamac_aur.db` ?

More importantly, try to run this:

`cd && find . -maxdepth 2 -type f -exec grep -Hn "sudo pamac" {} \;`

**Comment by FaFre** [comment]
Thank you for your help @jazz-it 

In the end I decided to remove `/var/tmp/pamac/` and `/var/tmp/pamac-build*` directories to give pamac a fresh start. A lot of things were messed up in this directories.

**Comment by jazz-it** [comment]
Sounds great. However, I advise you to run my last command, since it may help you indicate if there's any "sudo pamac" that slipped in somewhere, which may have caused all the mess (that may repeat again eventually). The same command helped me to identify my own culprit. For some reason I had `sudo pamac checkupdates -a` in my `~/utils` scripts, which caused all the trouble and brought me here. I had even `sudo pamac upgrade` in my aliases, which I stated in my comments above, which was probably another brainfart. Now that I removed `sudo` before `pamac`, it works smoothly again.

**Comment by jazz-it** [comment]
Hi @puxplaying, not sure if you had similar issues before, but perhaps it may be automated to avoid similar errors with faulty mirrors and switch to the next one in the line (if available)? Error example:

```
:: Synchronizing package databases...
 core                                                                    146,8 KiB   677 KiB/s 00:00 [------------------------------------------------------------] 100%
 extra                                                                     8,6 MiB  3,98 MiB/s 00:02 [------------------------------------------------------------] 100%
 multilib                                                                145,1 KiB   631 KiB/s 00:00 [------------------------------------------------------------] 100%
:: Starting full system upgrade...
resolving dependencies...
looking for conflicting packages...

Packages (1) firefox-120.0-1

Total Download Size:    64,45 MiB
Total Installed Size:  223,79 MiB
Net Upgrade Size:       -0,17 MiB

:: Proceed with installation? [Y/n] 
:: Retrieving packages...
 firefox-120.0-1-x86_64                                                   64,4 MiB  4,35 MiB/s 00:15 [------------------------------------------------------------] 100%
error: failed retrieving file 'firefox-120.0-1-x86_64.pkg.tar.zst' from manjaro.kurdy.org : The requested URL returned error: 404
(1/1) checking keys in keyring                                                                       [------------------------------------------------------------] 100%
(1/1) checking package integrity                                                                     [------------------------------------------------------------] 100%
(1/1) loading package files                                                                          [------------------------------------------------------------] 100%
(1/1) checking for file conflicts                                                                    [------------------------------------------------------------] 100%
(1/1) checking available disk space                                                                  [------------------------------------------------------------] 100%
:: Running pre-transaction hooks...
(1/1) Remove Firefox Distribution Settings
:: Processing package changes...
(1/1) upgrading firefox                                                                              [------------------------------------------------------------] 100%
:: Running post-transaction hooks...
(1/6) Arming ConditionNeedsUpdate...
(2/6) Copy Firefox Distribution Settings
(3/6) Updating icon theme caches...
(4/6) Checking for orphaned packages via hook...
=> No orphans found.
(5/6) Checking which packages need to be rebuilt
(6/6) Updating the desktop file MIME type cache...
```
My mirrors:
```
sudo pacman-mirrors --interactive --country Germany --api --protocol https;
sudo pacman-mirrors -f -g;
```

**Comment by puxplaying** [comment]
Hi @jazz-it , thanks for letting me know about the issue. As I never had this happen to me, I spent the last few days trying to reproduce it,  without success. 

The issue itself doesn't seem to come from pacman-mirrors directly, but rather from a database sync or download issue. So I will implement a small time-out after the mirror-sync in order to give it enough time to write the mirror-list even on the slowest HDD. Currently I've got no other indicator why this happens.

Is there any chance you are using a HDD?

**Comment by jazz-it** [comment]
> Is there any chance you are using a HDD?

My PC and my connection speed are quite fast and I use SSD M.2 NVMe for my Manjaro installation, not HDD. I reproduced it 2 days ago more than twice and assumed something was wrong (probably just temporarily) with the `manjaro.kurdy.org`, so I just unchecked it from my list of mirrors. Although the error showed up when I used the mirror above, surprisingly, all the installations (that have been affected) have been completed somehow. I didn't face similar issues since then, but would be great if your script could avoid such issues. Thanks!

**Comment by puxplaying** [comment]
> manjaro.kurdy.org

It seems this mirror is more troublesome than others, as other people on the [Manjaro forum](https://forum.manjaro.org/t/manjaro-kurdy-org-is-improperly-set-up/154900) are reporting about it not working. Basically the same issue that you had with it. Maybe your solution to exclude the mirror is currently the best advise to avoid this issue.

**Comment by puxplaying** [comment]
> Also, probably removing `--noconfirm` at lines 196 and 223 would be more safer to use?

Hi @jazz-it, this and more is now possible with the "-i" flag. You may want to give it a try. 😉
